8295724: VirtualMachineError: Out of space in CodeCache for method handle intrinsic

TheRealMDoerr opened the pull request:
This proposal prevents the VM from terminating unexpectedly in some rare cases (see JBS issue). It allows using NonNMethod code space for method handle intrinsics which are needed urgently if the other code cache spaces are full. There are other options (see JBS issue), but this one appears to be the simplest one.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295724](https://bugs.openjdk.org/browse/JDK-8295724): VirtualMachineError: Out of space in CodeCache for method handle intrinsic


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [2c5d2839](https://git.openjdk.org/jdk/pull/10933/files/2c5d2839038bc2cb2574d5eda580dc82ee7e8403)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10933/head:pull/10933` \
`$ git checkout pull/10933`

Update a local copy of the PR: \
`$ git checkout pull/10933` \
`$ git pull https://git.openjdk.org/jdk pull/10933/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10933`

View PR using the GUI difftool: \
`$ git pr show -t 10933`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10933.diff">https://git.openjdk.org/jdk/pull/10933.diff</a>

</details>

vnkozlov reviewed (commented):
Is it possible to test this case possibly with small CodeCache sizes?

TheRealMDoerr commented:
> Is it possible to test this case possibly with small CodeCache sizes?

We don't have any deterministic reproduction case. The problem only happens sometimes on some machines with a proprietary test, not always. For reproduction, both code cache segments need to be completely full and the GC needs to be slow enough at the point of time at which a new method handle intrinsic is required.
I only verified that the issue is gone with this change in our nightly tests.

vnkozlov reviewed (commented):
Thank you for testing explanation.

An other question: why only MH? May be this should be done for all native methods.

TheRealMDoerr commented:
> Thank you for testing explanation.
> 
> An other question: why only MH? May be this should be done for all native methods.

Only MH intrinsic allocation failures force the VM to terminate. Native wrappers are not required to continue execution. They can still get generated after GC has done its work and the other spaces are available again. I only wanted to use the NonNMethod space in emergency situations when the VM has no alternative. If we use it more often, it may get full, too, and then we wouldn't have any free space at all.

dean-long commented:
I'm not sure this change is safe.  We never put nmethods into NonNMethod.  If we did, a lot of things could break, because NMethodIterator doesn't look in that heap for nmethods.  We might get away with if method handle intrinsics don't have oops or inline caches, but it seems dangerous.  Also, this change only makes the failure harder to reproduce.  We could still run out of NonNMethod space, right?

dean-long commented:
I think we need a test for this case, even if it requires the use of WhiteBox to trigger it reliably.  For an alternative solution, can't we fall back to using the pre-generated interpreter entries?

TheRealMDoerr commented:
> I'm not sure this change is safe. We never put nmethods into NonNMethod (see code_blob_type_accepts_nmethod() and CodeCache::is_non_nmethod()). If we did, a lot of things could break, because NMethodIterator doesn't look in that heap for nmethods. We might get away with it if method handle intrinsics don't have oops or inline caches, but it seems dangerous. Also, this change only makes the failure harder to reproduce. We could still run out of NonNMethod space, right?

I think the method handle intrinsics were just implemented as nmethods because it was easy to do. They just perform method handle dispatch and contain no immediate oops. They are more like stubs (and like the interpreter versions which aren't nmethods, either). So, I can't see what could break by treating them this way.
To your 2nd question: No, we can't just fall back to using interpreter entries when calling from compiled code. The calling convention is not compatible.
We could get out of space in the NonNMethod segment, too, but the GC has more time until then. I have thought about waiting for the GC, but how long should we wait? A complete marking cycle of a huge Java heap may take time. Terminating the VM may still be undesired, but an unresponsive application, too. The situation sounds similar to "GC overhead limit exceeded".

dean-long commented:
I believe the calling convention issues would be handled by the c2i adapter.  We already force calling from the interpreter to use the i2c adapter.  It looks like the problem with falling back to interpreter intrinsics is _linkToNative, which is missing.  All platforms only generate it for compiled.  From code comments, it seems like this was an implementation short-cut and falls under technical debt.  Here is my suggestion:
1. File an RFE to support _linkToNative interpreter intrinsic, fallback to interpreted through c2i adapter
2. Proceed with this PR as a stop-gap
For this PR, it would be reassuring to assert that any nmethods added to NonNMethod have no oops and no inlined caches, which probably also implies no metadata and no relocations.  Also, I would really like a test that reproduces the problem.  Would using -XX:-MethodFlushing -XX:-UseCodeCacheFlushing help with that?

reinrich commented:
I think the continuation intrinsics (see `Method::is_continuation_native_intrinsic()`) are problematic too. There are no interpreted versions of them. It will be difficult to implement interpreted versions.

dean-long commented:
> I think the continuation intrinsics (see `Method::is_continuation_native_intrinsic()`) are problematic too. There are no interpreted versions of them. It will be difficult to implement interpreted versions.

Good point.  It's not really interpreted vs compiled that is important, it's eager-allocated vs lazy-allocated, and if we need a different one for every polymorphic signature.  Somewhere in between lazy and eager is "delayed".  We could chose a low-water mark, where the code cache has free space less than  "LazyIntrinsicsSize" available, where we force these intrinsics to get generated if they weren't already generated lazily.  The problem is the missing signature-independent interpreted _linkToNative and the fact that the compiled version depends on the polymorphic signature, forcing them to be generated lazily to handle all possible signatures.

TheRealMDoerr commented:
Thanks for all your interesting thoughts. But now, I think we're discussing too many things in one PR.
I suggest the following:
- New issue for the Loom intrinsics. We just need 2 native functions, so they could just get created at startup if there's a concern.
- New issue for C2I investigation. We may need to generate adapters in the NonNMethod space. Isn't that a similar problem?
- Discuss further if this PR is a viable solution for the method handle intrinsic issue in JDK 20. They get created per signature, so it's hard to generate all ones :-) I wouldn't expect an insane amount, so a bit of NonNMethod space can hopefully take as much as needed. If there are more concerns, we could still wait for GC.
- I'd like to try implementing a reproduction case, but I can't promise if I can get that done before RDP1. There's a lot to do atm. Could we add it later if it takes too long?

dean-long commented:
Those are reasonable suggestions.  Sounds like a plan.  For this PR, could you refactor the oops checking logic into its own method?  Then would it make sense to add checks for relocations that use metadata?  Some types of call sites are resolved lazily, so they could be using metadata without that being reflected in the oops.

TheRealMDoerr commented:
Filed 2 issues:
https://bugs.openjdk.org/browse/JDK-8296336
https://bugs.openjdk.org/browse/JDK-8296334
Feel free to modify them.

I've added a check for metadata. I don't think we need to check for metadata relocations because the oop recorder always records all metadata pointers. We never iterate over the code to find metadata AFAIK. Please take a look.

iklam commented:
Is it possible to have a backup stub? It can trap to c code, do what the method handle intrinsics is supposed to do (by iterating on the signature), and continue execution.

dean-long commented:
> Filed 2 issues: https://bugs.openjdk.org/browse/JDK-8296336 https://bugs.openjdk.org/browse/JDK-8296334 Feel free to modify them.
> 
> I've added a check for metadata. I don't think we need to check for metadata relocations because the oop recorder always records all metadata pointers. We never iterate over the code to find metadata AFAIK (except for patching purposes, but the metadata is duplicated in this case, not in the code only). Please take a look.

The metadata in stubs might not have an oop.  See the comment here:
https://github.com/openjdk/jdk/blob/054c23f484522881a0879176383d970a8de41201/src/hotspot/share/code/compiledMethod.cpp#L632

dean-long commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -474,13 +474,23 @@ nmethod* nmethod::new_native_nmethod(const methodHandle& method,
             basic_lock_owner_sp_offset,
             basic_lock_sp_offset,
             oop_maps);
+#ifdef ASSERT
+    // Method handle instrinsics may get allocated in NonNMethod space.
+    // GC may not look for Oops, there.
+    if (method->is_method_handle_intrinsic()) {
+      assert(oop_maps == nullptr, "expectation");
+      assert(!nm->contains_oops(), "no oops allowed");
+      assert(nm->metadata_size() == 0, "metadata usage not expected");
+    }
+#endif
```
It might make sense to do this check for all nmethods that were allocated in NonNMethod.

dean-long commented:
> Is it possible to have a backup stub? It can trap to c code, do what the method handle intrinsics is supposed to do (by iterating on the signature), and continue execution.

Yes, I was thinking the interpreter _linkToNative might need to do that, similar to how old versions of the JDK had a generic native method wrapper that could handle all signatures.  That can be investigated as part of JDK-8296336.

TheRealMDoerr commented:
> > Filed 2 issues: https://bugs.openjdk.org/browse/JDK-8296336 https://bugs.openjdk.org/browse/JDK-8296334 Feel free to modify them.
> > I've added a check for metadata. I don't think we need to check for metadata relocations because the oop recorder always records all metadata pointers. We never iterate over the code to find metadata AFAIK (except for patching purposes, but the metadata is duplicated in this case, not in the code only). Please take a look.
> 
> The metadata in stubs might not have an oop. See the comment here:
> 
> https://github.com/openjdk/jdk/blob/054c23f484522881a0879176383d970a8de41201/src/hotspot/share/code/compiledMethod.cpp#L632

I think this code was written for x86 and is based on a wrong assumption. If a static call (or optimized virtual call) is reachable then the callee class can’t be dead. Unfortunately, JDK-8222841 didn’t include a test. I’ve removed the `CompiledMethod::cleanup_inline_caches_impl` change and ran a few tests like `vmTestbase/nsk/jvmti/RedefineClasses/StressRedefine/TestDescription.java`. Didn’t show errors.
Other platforms (e.g. PPC64, AArch64) do allocate a metadata slot via oop recorder. The code you are referring to only clears the slot in the metadata section, not the replicated value which is actually used by the code. Note that `fix_metadata_relocation()` does nothing on most platforms! This is a further indication that the code is not needed.
Should I file another RFE for that?

dean-long commented:
> This is a further indication that the code is not needed. Should I file another RFE for that?

Please do, and link JDK-8294002 as related.  I suspected that the duplicated metadata on aarch64 could be a problem, if we ended up with different values, however I wasn't able to reproduce a problem.

For this PR, would it make sense to use a MetadataClosure and metadata_do to check for metadata?

TheRealMDoerr commented:
> > This is a further indication that the code is not needed. Should I file another RFE for that?
> 
> Please do, and link JDK-8294002 as related. I suspected that the duplicated metadata on aarch64 could be a problem, if we ended up with different values, however I wasn't able to reproduce a problem.
> 
> For this PR, would it make sense to use a MetadataClosure and metadata_do to check for metadata?

New Issue:
https://bugs.openjdk.org/browse/JDK-8296440
Regarding the aarch64 question: I don't think we ever hit the case in which the Method* gets cleared. So, both values never differ.

It is possible to reuse the existing code (oops_do and metadata_do). Please see my latest commit. Requires extra walks in debug build, but I think it's affordable.

dean-long commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -474,13 +490,30 @@ nmethod* nmethod::new_native_nmethod(const methodHandle& method,
             basic_lock_owner_sp_offset,
             basic_lock_sp_offset,
             oop_maps);
+#ifdef ASSERT
```
How about making this whole block a separate function?

TheRealMDoerr replied:
Done. And thanks for the hint regarding `is_permanent_class_loader_data`. That was what I was missing. Assertion added.

dean-long commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -474,13 +490,30 @@ nmethod* nmethod::new_native_nmethod(const methodHandle& method,
+      CheckForOopsClosure cfo;
+      nm->oops_do(&cfo);
+      assert(!cfo.found_oop(), "no oops allowed");
+      CheckForMetadataClosure cfm(/* ignore reference to own Method */ nm->method());
```
If we are going to ignore method(), then how about checking that it's classloader is parmanent (ClassLoaderData::is_permanent_class_loader_data)?  We also might need a review by a GC expert.  @fisk, do you agree it is safe to put nmethod in NonNMethod space given the above checks?

fisk replied:
Hmm. The GC used to walk all nmethods an ensure the is_unloading_state is computed for all of them. When using a STW collector, we will crash if trying to compute this outside of the UnloadingScope, as we then don't have rnough information to be able to compute it. So I suppose this is a bit fragile; if an nmethod is acquired from outside of an nmethod iterator, and the is_unloading() question is asked, we risk crashing.
I suppose we could strengthen this by letting is_unloading() know rhat the answer is false for method handle intrinsics.
It feels more and more like method handle intrinsics shouldn't be nmethods as more and more code that deals with nmethods needs to know that this isn't really an nmethod. Although that's not a new observation for this patch.

TheRealMDoerr replied:
So, calling `IsUnloadingBehaviour::is_unloading` is unsafe. Should we only call it when `is_permanent_class_loader_data()` for the method returns false?

dean-long replied:
Does this also mean that nmethods for method handle intrinsics could previously get unloaded when they became "cold", and now if they are in NonNMethod they live forever?  Someone could probably write a test that fills up NonNMethod with method handle intrinsics for various signatures by spinning up temporary classes with random signatures.

TheRealMDoerr replied:
We were able to hit the `IsUnloadingBehaviour::is_unloading` problem! I'm testing a workaround, now.
Note that the Methods are also permanently kept in `_invoke_method_intrinsic_table`. Getting rid of method handle adapters is currently not supported.

dean-long commented:
It sounds like using nmethods for compiled method handle intrinsics isn't necessary and causes problems.  What if we store them as MethodHandlesAdapterBlobs like we do for the interpreter?  Lookup would need to go through method->adapter() instead of method->code().  Is there already an RFE filed for this?

TheRealMDoerr commented:
> It sounds like using nmethods for compiled method handle intrinsics isn't necessary and causes problems. What if we store them as MethodHandlesAdapterBlobs like we do for the interpreter? Lookup would need to go through method->adapter() instead of method->code(). Is there already an RFE filed for this?

This may be a better approach. Generation should probably get moved out of `generate_native_wrapper`. I'll take a closer look when I find more time.

TheRealMDoerr commented:
I think treating method handle intrinsics as adapter blobs has other drawbacks. At least, we would need to think about how much space to reserve in non-nmethod space. What if that one gets full?
I think https://bugs.openjdk.org/browse/JDK-8296336 is the best approach in the long term.
For now, this PR has worked stable for 2 weeks.
By the way, I was able to reproduce the issue by `jdk/bin/java -XX:ReservedCodeCacheSize=2496k -XX:+SegmentedCodeCache -version`. But, I'll take a look at the whitebox API and see if I can write a more deterministic test.

TheRealMDoerr commented:
Thanks for the reviews and all the helpful comments and ideas! I'm planning to integrate before JDK 20 rampdown once the tests have passed. Our internal testing didn't find any further problems.

vnkozlov commented:
New test failed on all platforms when run with '-ea -esa -XX:CompileThreshold=100 -XX:-TieredCompilation'

```
----------System.out:(11/726)----------
CompileCommand: compileonly null.* bool compileonly = true
[0.427s][warning][codecache] CodeHeap 'non-profiled nmethods' is full. Compiler has been disabled.
[0.427s][warning][codecache] Try increasing the code heap size using -XX:NonProfiledCodeHeapSize=
CodeHeap 'non-profiled nmethods': size=11236Kb used=11203Kb max_used=11203Kb free=33Kb
 bounds [0x00007ff0bd7f9000, 0x00007ff0be2f2000, 0x00007ff0be2f2000]
CodeHeap 'non-nmethods': size=5148Kb used=1630Kb max_used=1630Kb free=3517Kb
 bounds [0x00007ff0bd2f2000, 0x00007ff0bd562000, 0x00007ff0bd7f9000]
 total_blobs=932 nmethods=51 adapters=711
 compilation: disabled (not enough contiguous free space left)
              stopped_count=1, restarted_count=0
 full_count=1
----------System.err:(14/1222)----------
Java HotSpot(TM) 64-Bit Server VM warning: CodeHeap 'non-profiled nmethods' is full. Compiler has been disabled.
Java HotSpot(TM) 64-Bit Server VM warning: Try increasing the code heap size using -XX:NonProfiledCodeHeapSize=
java.lang.NullPointerException: Cannot invoke "java.lang.management.MemoryPoolMXBean.getUsage()" because "<local1>" is null
	at compiler.codecache.MHIntrinsicAllocFailureTest.fillCodeCacheSegment(MHIntrinsicAllocFailureTest.java:60)
	at compiler.codecache.MHIntrinsicAllocFailureTest.main(MHIntrinsicAllocFailureTest.java:70)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
	at java.base/java.lang.Thread.run(Thread.java:1599)
```

TobiHartmann commented:
I re-submitted testing and will report back once it passed.

TheRealMDoerr commented:
Thanks for testing!
/integrate
